### ISSUE TITLE
Does not restore behavior correctly after orientation change

### ISSUE THREAD

**Issue by overengineered** [henrytao-me/smooth-app-bar-layout#61]
Steps to reproduce:
1. Start activity with SmoothAppBarLayout, CollapsingToolbarLayout with `app:layout_scrollFlags="scroll|exitUntilCollapsed"` and a long list of items below
2. Scroll to the bottom of list
3. Rotate phone to the side
4. Rotate phone back
5. verify that list bottom is displayed (unfortunately this cannot be reproduced in Sample application as it does not preserve list position between orientation changes)
6. verify that toolbar is show collapsed as it should (we are using exitUntilCollapsed)
7. scroll up a little bit

What happens:
  Toolbar is suddenly shown in expanded state.

What should happen:
  Toolbar should stay collapsed. It should only expand once list is scrolled all the way to the top.

Workaround:
1. Add a class in the app
   
   ```
   public class PersistentAppBarLayoutBehavior extends me.henrytao.smoothappbarlayout.SmoothAppBarLayout.Behavior
   {
   public PersistentAppBarLayoutBehavior()
   {
       super();
   }
   
   public PersistentAppBarLayoutBehavior(Context context, AttributeSet attrs)
   {
       super(context, attrs);
   }
   
   @Override
   public Parcelable onSaveInstanceState(CoordinatorLayout parent, AppBarLayout appBarLayout)
   {
       Bundle behaviorState = new Bundle();
       behaviorState.putInt("mCurrentScrollOffset", mCurrentScrollOffset);
       behaviorState.putInt("mCurrentTranslationOffset", mCurrentTranslationOffset);
       behaviorState.putInt("mQuickReturnOffset", mQuickReturnOffset);
   
       behaviorState.putParcelable("super", super.onSaveInstanceState(parent, appBarLayout));
   
       return behaviorState;
   }
   
   @Override
   public void onRestoreInstanceState(CoordinatorLayout parent, AppBarLayout appBarLayout,
                                      Parcelable state)
   {
       Bundle behaviorState = (Bundle) state;
       behaviorState.setClassLoader(appBarLayout.getContext().getClassLoader());
   
       mCurrentScrollOffset = behaviorState.getInt("mCurrentScrollOffset");
       mCurrentTranslationOffset = behaviorState.getInt("mCurrentTranslationOffset");
       mQuickReturnOffset = behaviorState.getInt("mQuickReturnOffset");
   
       Parcelable superState = behaviorState.getParcelable("super");
   
       super.onRestoreInstanceState(parent, appBarLayout, superState);
   }
   }
   ```
2. change me.henrytao.smoothappbarlayout.SmoothAppBarLayout behavior `<me.henrytao.smoothappbarlayout.SmoothAppBarLayout
   android:id="@+id/layoutAppBar"
   app:layout_behavior="your.package.PersistentAppBarLayoutBehavior">`
3. you may need to add [ids to all the elements](http://stackoverflow.com/questions/32563300/) in the layout

EDITED 2016-02-09: small fix in workaround

**Comment by henrytao-me** [comment]
Hi @overengineered 

Thanks for your feedback. I will fix it in up coming version. 

Cheers, 
Henry

**Comment by bachphuc** [comment]
Hi.
First, thanks for great library. I'm trying to integrate your library for my app.
I have same issue like above.

My Java code

``` java
package me.henrytao.smoothappbarlayoutdemo.activity;

import android.app.Activity;
import android.os.Bundle;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;

import butterknife.Bind;
import butterknife.ButterKnife;
import me.henrytao.smoothappbarlayout.PagerAdapter;
import me.henrytao.smoothappbarlayoutdemo.R;
import me.henrytao.smoothappbarlayoutdemo.apdater.ViewPagerRunnableAdapter;
import me.henrytao.smoothappbarlayoutdemo.fragment.DummyNestedScrollViewFragment;
import me.henrytao.smoothappbarlayoutdemo.fragment.DummyRecyclerViewFragment;
import me.henrytao.smoothappbarlayoutdemo.fragment.TestFragment;

public class TestActivity extends AppCompatActivity {
    @Bind(me.henrytao.smoothappbarlayoutdemo.R.id.collapsing_toolbar_layout)
    CollapsingToolbarLayout vCollapsingToolbarLayout;

    @Bind(me.henrytao.smoothappbarlayoutdemo.R.id.toolbar)
    Toolbar vToolbar;

    @Bind(me.henrytao.smoothappbarlayoutdemo.R.id.tab_layout)
    TabLayout vTabLayout;

    @Bind(me.henrytao.smoothappbarlayoutdemo.R.id.view_pager)
    ViewPager vViewPager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // setContentView(R.layout.activity_test2);
        setContentView(R.layout.activity_test);

        /*Fragment fragment = new TestFragment();
        if (fragment != null) {
            getSupportFragmentManager()
                    .beginTransaction()
                    .replace(R.id.fragment, fragment)
                    .commit();
        }*/

        ButterKnife.bind(this);

        ViewPagerRunnableAdapter adapter = new ViewPagerRunnableAdapter(getSupportFragmentManager());
        adapter.addFragment(DummyRecyclerViewFragment.newInstance("Cat", 100, me.henrytao.smoothappbarlayoutdemo.R.layout.item_header_spacing), "Cat");
        adapter.addFragment(DummyRecyclerViewFragment.newInstance("Dog", 100, me.henrytao.smoothappbarlayoutdemo.R.layout.item_header_spacing), "Dog");
        adapter.addFragment(DummyRecyclerViewFragment.newInstance("Mouse", 100, me.henrytao.smoothappbarlayoutdemo.R.layout.item_header_spacing), "Mouse");
        adapter.addFragment(DummyRecyclerViewFragment.newInstance("Chicken", 5, me.henrytao.smoothappbarlayoutdemo.R.layout.item_header_spacing), "Chicken");
        adapter.addFragment(DummyNestedScrollViewFragment.newInstance(getString(me.henrytao.smoothappbarlayoutdemo.R.string.text_long),
                me.henrytao.smoothappbarlayoutdemo.R.layout.item_header_spacing), "Duck");
        adapter.addFragment(DummyRecyclerViewFragment.newInstance("Bird", 100, me.henrytao.smoothappbarlayoutdemo.R.layout.item_header_spacing), "Bird");
        adapter.addFragment(DummyNestedScrollViewFragment.newInstance(getString(me.henrytao.smoothappbarlayoutdemo.R.string.text_short),
                me.henrytao.smoothappbarlayoutdemo.R.layout.item_header_spacing), "Tiger");

        // PagerAdapter have to implement `me.henrytao.smoothappbarlayout.PagerAdapter` in order to make it work with `SmoothAppBarLayout`
        if (adapter instanceof PagerAdapter) {
            vViewPager.setAdapter(adapter);
        }
        vTabLayout.setupWithViewPager(vViewPager);
        vTabLayout.setTabMode(TabLayout.MODE_SCROLLABLE);

        setSupportActionBar(vToolbar);
    }
}
```

and xml code

``` xml
<android.support.design.widget.CoordinatorLayout xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto"
    android:layout_width="match_parent"
    android:layout_height="match_parent">

    <android.support.v4.view.ViewPager
        android:id="@+id/view_pager"
        android:layout_width="match_parent"
        android:layout_height="match_parent" />

    <me.henrytao.smoothappbarlayout.SmoothAppBarLayout
        android:layout_width="match_parent"
        android:layout_height="@dimen/app_bar_height">

        <android.support.design.widget.CollapsingToolbarLayout
            android:id="@+id/collapsing_toolbar_layout"
            android:layout_width="match_parent"
            android:layout_height="match_parent"
            app:contentScrim="?attr/colorPrimary"
            app:layout_scrollFlags="scroll|exitUntilCollapsed">

            <ImageView
                android:id="@+id/cover"
                android:layout_width="match_parent"
                android:layout_height="match_parent"
                android:background="@drawable/cover"
                android:scaleType="centerCrop"
                app:layout_collapseMode="parallax" />

            <android.support.v7.widget.Toolbar
                android:id="@+id/toolbar"
                style="@style/AppStyle.MdToolbar"
                android:layout_height="104dp"
                android:background="@android:color/transparent"
                android:gravity="top"
                android:minHeight="?attr/actionBarSize"
                app:layout_collapseMode="pin"
                app:navigationIcon="@drawable/ic_toolbar_arrow_back"
                app:titleMarginTop="13dp" />

            <android.support.design.widget.TabLayout
                android:id="@+id/tab_layout"
                android:layout_width="match_parent"
                android:layout_height="wrap_content"
                android:layout_gravity="bottom" />
        </android.support.design.widget.CollapsingToolbarLayout>
    </me.henrytao.smoothappbarlayout.SmoothAppBarLayout>
</android.support.design.widget.CoordinatorLayout>
```

It's working if orientation not change, when orientation change app bar not scroll, it alway expand.
Lets me know if I was wrong.

Thanks
Regards

**Comment by henrytao-me** [comment]
As @overengineered proposed, it needs to save and restore some parameters. Let me work on that and ping you when it's ready (probably soon). 

Cheers,
Henry

**Comment by bachphuc** [comment]
Thanks you.

**Comment by petersnoopy** [comment]
+1

**Comment by henrytao-me** [comment]
Hi @overengineered 

I fixed this issue in v1.0.3. Check out new implementation there. I will write some documentations later. 
Thanks a lot. 

Cheers, 
Henry
